routes/post.go

```diff
@@ -1203,13 +1191,153 @@ func (fes *APIServer) GetSinglePost(ww http.ResponseWriter, req *http.Request) {
+			profilePubKeyMap[pkMapKey] = commentEntry.PosterPublicKey
+		}
+	}
+
+	// Filter out restricted PosterPublicKeys.
+	filteredProfilePubKeyMap, err := fes.FilterOutRestrictedPubKeysFromMap(
+		profilePubKeyMap, readerPublicKeyBytes, "leaderboard" /*moderationType*/, utxoView)
+	if err != nil {
+		return nil, err
+	}
+
+
+	// Figure out if the current poster is greylisted.  If the current poster is greylisted, we will add their
+	// public key back to the filteredProfileMap so their profile will appear for the current post
+	// and any parents, but we will remove any comments by this greylisted user.
+	isCurrentPosterGreylisted := false
+	if _, ok := filteredProfilePubKeyMap[lib.MakePkMapKey(posterPublicKeyBytes)]; !ok {
+		// Get the userMetadata for the currentPosters
+		currentPosterUserMetadataKey := append([]byte{}, _GlobalStatePrefixPublicKeyToUserMetadata...)
+		currentPosterUserMetadataKey = append(currentPosterUserMetadataKey, posterPublicKeyBytes...)
+		var currentPosterUserMetadataBytes []byte
+		currentPosterUserMetadataBytes, err = fes.GlobalState.Get(currentPosterUserMetadataKey)
+		if err != nil {
+			return nil, err
+		}
+		// If the currentPoster's userMetadata doesn't exist, then they are no greylisted, so we can exit.
+		if currentPosterUserMetadataBytes != nil {
+			// Decode the currentPoster's userMetadata.
+			currentPosterUserMetadata := UserMetadata{}
+			err = gob.NewDecoder(bytes.NewReader(currentPosterUserMetadataBytes)).Decode(&currentPosterUserMetadata)
+			if err != nil {
+				return nil, err
+			}
+			// If the currentPoster is not blacklisted (removed everywhere) and is greylisted (removed from leaderboard)
+			// add them back to the filteredProfilePubKeyMap and note that the currentPoster is greylisted.
+			if currentPosterUserMetadata.RemoveFromLeaderboard && !currentPosterUserMetadata.RemoveEverywhere {
+				isCurrentPosterGreylisted = true
+				filteredProfilePubKeyMap[lib.MakePkMapKey(posterPublicKeyBytes)] = posterPublicKeyBytes
+			}
+		}
+	}
+
+	// If the profile that posted this post is not in our filtered list, return with error.
+	if filteredProfilePubKeyMap[lib.MakePkMapKey(posterPublicKeyBytes)] == nil && !isCurrentPosterGreylisted {
+		return nil, err
+	}
+
+	// Get the profile entry for all PosterPublicKeys that passed our filter.
+	pubKeyToProfileEntryResponseMap := make(map[lib.PkMapKey]*ProfileEntryResponse)
+	for _, pubKeyBytes := range filteredProfilePubKeyMap {
+		profileEntry := utxoView.GetProfileEntryForPublicKey(pubKeyBytes)
+		if profileEntry == nil {
+			continue
+		} else {
+			pubKeyToProfileEntryResponseMap[lib.MakePkMapKey(pubKeyBytes)] =
+				fes._profileEntryToResponse(profileEntry, utxoView)
+		}
+	}
+
+	// If the profile that posted this post does not have a profile, return with error.
+	if pubKeyToProfileEntryResponseMap[lib.MakePkMapKey(posterPublicKeyBytes)] == nil {
```

Comment:
let save the PkMapKey for poster up above. It's a relatively inexpensive operation, but we really don't need to keep creating it.